### ISSUE TITLE
Bug: createIsolatedProgram fallback incorrectly triggered with projectService when external parser calls parseAndGenerateServices multiple times

### ISSUE THREAD

**Issue by ryo-chin** [typescript-eslint/typescript-eslint#12065]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

When using `projectService: true` with `vue-eslint-parser` v9 in a CI environment (where `singleRun` is inferred as `true`), type-aware lint rules incorrectly report that `strictNullChecks` is not enabled, even though `strict: true` is set in `tsconfig.json`.

**Root cause:** `vue-eslint-parser` calls `@typescript-eslint/parser`'s `parseAndGenerateServices()` multiple times for the same file path — once for the `<script>` block and additional times for template expressions (e.g., `{{ x }}`, `v-if`, `v-slot`). This has been the case since well before v9 and is a fundamental part of how `vue-eslint-parser` handles template expression linting. In `singleRun` mode, the second call increments `parseAndGenerateServicesCalls` above 1, which triggers the `createIsolatedProgram` fallback ([`parser.ts` L234-L239](https://github.com/typescript-eslint/typescript-eslint/blob/68a074fb6b4a91f04c4bbbd42239c22aec498e78/packages/typescript-estree/src/parser.ts#L234-L239)).

The current fallback logic assumes that "same file parsed more than once in singleRun = ESLint `--fix` cycle". However, it may also be worth considering the case where external parsers like `vue-eslint-parser` legitimately call the parser multiple times for the same file during a single lint pass (for template expressions, etc.).

`createIsolatedProgram` does not read `tsconfig.json`, so all compiler options (including `strictNullChecks`) fall back to TypeScript defaults (`false`), breaking type-aware rules.

**Regression:** This behavior has been latent for a long time, but was never triggered because `inferSingleRun` unconditionally returned `false` when `extraFileExtensions` was set — effectively preventing `singleRun=true` for all `.vue` users. PR #11327 (v8.48.0, November 2025) changed this guard to apply only to the `project` option, allowing `projectService` users to enter `singleRun=true` mode and exposing the pre-existing incompatibility.

### Reproduction Repository Link

https://github.com/ryo-chin/typescript-eslint-issue-repro

### Repro Steps

1. Clone the repo
2. `npm install`
3. `CI=true npx eslint --no-cache src/App.vue`

**Expected:** Rule reports actual code issues (e.g., "Unexpected nullish value in conditional").

**Actual:** Rule reports `This rule requires the 'strictNullChecks' compiler option to be turned on to function correctly`.

**Verification:** Running with `TSESTREE_SINGLE_RUN=false npx eslint --no-cache src/App.vue` produces the correct result.

### Suggested Fix

Skip the `createIsolatedProgram` fallback when `projectService` is in use:

```diff
  const { ast, program } =
    parseSettings.singleRun &&
    tsestreeOptions.filePath &&
-   parseAndGenerateServicesCalls[tsestreeOptions.filePath] > 1
+   parseAndGenerateServicesCalls[tsestreeOptions.filePath] > 1 &&
+   !parseSettings.projectService
      ? createIsolatedProgram(parseSettings)
      : getProgramAndAST(parseSettings, hasFullTypeInformation);
```

This fallback exists for the legacy `project` option, which uses AOT-compiled programs that can become stale during `--fix` cycles. `projectService` does not use AOT programs — it passes the latest source code to `openClientFile()` on every call ([`useProgramFromProjectService.ts` L150-L156](https://github.com/typescript-eslint/typescript-eslint/blob/68a074fb6b4a91f04c4bbbd42239c22aec498e78/packages/typescript-estree/src/useProgramFromProjectService.ts#L150-L156)), so no fallback is necessary. Verified that `projectService` + `singleRun=true` + `--fix` works correctly without the fallback.

### Scope and Impact

This reproduction uses `vue-eslint-parser` v9, but the multiple-call behavior for template expressions has been present in `vue-eslint-parser` since well before v9. It was never an issue because the `extraFileExtensions` guard in `inferSingleRun` kept `.vue` users in `singleRun=false` mode, preventing the fallback from being triggered.

`vue-eslint-parser` v10 (released March 2025) optimized its internals to call the parser only once per file, so projects using v10 are not affected. Since `typescript-eslint` v8.48.0 (which introduced this regression) was released in November 2025, most actively maintained Vue projects had likely already migrated to `eslint-plugin-vue` v10 / `vue-eslint-parser` v10 by that time. This may explain why no issues have been filed so far — the affected combination (`vue-eslint-parser` v9 + `projectService` + CI) is relatively uncommon at this point.

That said, the underlying issue is not specific to `vue-eslint-parser`. The `createIsolatedProgram` fallback can be triggered by any external parser or plugin that calls `parseAndGenerateServices()` multiple times for the same file path in a single lint pass. A similar issue was previously reported with `eslint-plugin-import` (#5880 / #5136), where out-of-band parsing incremented the call counter and triggered the fallback. That case was fixed in PR #5893 by preventing singleRun mode when not in type-aware linting. However, today's case differs: the multiple calls happen *during* type-aware linting (template expression parsing), so that guard does not apply. Fixing this in `typescript-estree` for the `projectService` case would make the fallback logic more robust against this class of interaction.

### Workarounds

- Set `disallowAutomaticSingleRunInference: true` in `parserOptions`
- Set environment variable `TSESTREE_SINGLE_RUN=false` in CI
- Upgrade `vue-eslint-parser` to v10 (which calls the parser only once per file)

### Related Issues

- #11327 — PR that exposed this bug by allowing `projectService` users into `singleRun=true` mode
- #5880 / #5136 — Same class of bug: `eslint-plugin-import`'s out-of-band parsing incremented the call counter, triggering `createIsolatedProgram`. Fixed in PR #5893 by preventing singleRun mode when not in type-aware linting. Today's case differs: the multiple calls happen *during* type-aware linting, so that guard does not help.
- #9344 — Same class of bug: `createIsolatedProgram` triggered during `--fix` cycles
- #5856 — Refactor issue noting `createIsolatedProgram` fallback "should never be hit"

### Versions

| package                                | version   |
| -------------------------------------- | --------- |
| `@typescript-eslint/eslint-plugin`     | `8.55.0`  |
| `@typescript-eslint/parser`            | `8.55.0`  |
| `@typescript-eslint/typescript-estree` | `8.55.0`  |
| `TypeScript`                           | `5.9.3`   |
| `ESLint`                               | `10.0.0`  |
| `node`                                 | `24.7.0`  |

**Comment by darshjme-codes** [comment]
**Triage note:**

Excellent bug report! This is a regression introduced in v8.48.0 (PR #11327) where `projectService` users can incorrectly trigger `createIsolatedProgram` fallback in CI environments.

**Root cause:** vue-eslint-parser legitimately calls `parseAndGenerateServices()` multiple times for the same file (once for `<script>`, additional times for template expressions), which increments the call counter and triggers the fallback in singleRun mode. `createIsolatedProgram` doesn't read tsconfig.json, breaking type-aware rules.

**Suggested labels:**
- `bug` ✓
- `package: typescript-estree`
- `regression`
- `affects: vue-eslint-parser`

**Impact:** Affects projectService users with vue-eslint-parser v9 in CI (relatively rare since v10 is widely adopted).

**Proposed fix in issue is sound:** Skip `createIsolatedProgram` fallback when `projectService` is enabled, since projectService doesn't use AOT programs and always has latest source code.

cc @ryo-chin - Great analysis and suggested fix! 🎯
